src/split.cc

```diff
@@ -0,0 +1,155 @@
+#include "split.h"
+
+#include <array>
+#include <tuple>
+
+#include "grid.h"
+
+namespace {
+
+using namespace celerity;
+using namespace celerity::detail;
+
+void sanity_check_split(const chunk<3>& full_chunk, const std::vector<chunk<3>>& split) {
+	region<3> reconstructed_chunk;
+	for(auto& chnk : split) {
+		assert(region_intersection(reconstructed_chunk, box<3>(chnk)).empty());
+		reconstructed_chunk = region_union(box<3>(chnk), reconstructed_chunk);
+	}
+	assert(region_difference(reconstructed_chunk, box<3>(full_chunk)).empty());
+}
+
+template <int Dims>
+std::tuple<std::array<size_t, Dims>, std::array<size_t, Dims>, std::array<size_t, Dims>> compute_small_and_large_chunks(
+    const chunk<3>& full_chunk, const range<3>& granularity, const std::array<size_t, Dims>& actual_num_chunks) {
+	std::array<size_t, Dims> small_chunk_size{};
+	std::array<size_t, Dims> large_chunk_size{};
+	std::array<size_t, Dims> num_large_chunks{};
+	for(int d = 0; d < Dims; ++d) {
+		const size_t ideal_chunk_size = full_chunk.range[d] / actual_num_chunks[d];
+		small_chunk_size[d] = (ideal_chunk_size / granularity[d]) * granularity[d];
+		large_chunk_size[d] = small_chunk_size[d] + granularity[d];
+		num_large_chunks[d] = (full_chunk.range[d] - small_chunk_size[d] * actual_num_chunks[d]) / granularity[d];
+	}
+	return {small_chunk_size, large_chunk_size, num_large_chunks};
+}
+
+} // namespace
+
+namespace celerity::detail {
+
+std::vector<chunk<3>> split_1d(const chunk<3>& full_chunk, const range<3>& granularity, const size_t num_chunks) {
+#ifndef NDEBUG
+	assert(num_chunks > 0);
+	for(int d = 0; d < 3; ++d) {
+		assert(granularity[d] > 0);
+		assert(full_chunk.range[d] % granularity[d] == 0);
+	}
+#endif
+
+	// Due to split granularity requirements or if num_workers > global_size[0],
+	// we may not be able to create the requested number of chunks.
+	const std::array<size_t, 1> actual_num_chunks = {std::min(num_chunks, full_chunk.range[0] / granularity[0])};
+	const auto [small_chunk_size, large_chunk_size, num_large_chunks] = compute_small_and_large_chunks<1>(full_chunk, granularity, actual_num_chunks);
+
+	std::vector<chunk<3>> result(actual_num_chunks[0], {full_chunk.offset, full_chunk.range, full_chunk.global_size});
+	for(auto i = 0u; i < num_large_chunks[0]; ++i) {
+		result[i].range[0] = large_chunk_size[0];
+		result[i].offset[0] += i * large_chunk_size[0];
+	}
+	for(auto i = num_large_chunks[0]; i < actual_num_chunks[0]; ++i) {
+		result[i].range[0] = small_chunk_size[0];
+		result[i].offset[0] += num_large_chunks[0] * large_chunk_size[0] + (i - num_large_chunks[0]) * small_chunk_size[0];
+	}
+
+#ifndef NDEBUG
+	sanity_check_split(full_chunk, result);
+#endif
```

Comment:
I appreciate this. (And it's another good reason that we should emphasize the utility of debug -- or optimized-with-debug-features -- builds for development in the documentation in the future)